### ISSUE TITLE
Neuron state initialization 

### ISSUE THREAD

**Issue by asudeeaydin** [fangwei123456/spikingjelly#263]
Hi there,

is it possible to initialize the states of neurons to anything other than 0? 
I'm suing spiking jelly version 0.0.0.0.6.

**Comment by fangwei123456** [comment]
https://github.com/fangwei123456/spikingjelly/blob/cb2f0b9fb0659b1ddfa320ea83434cfa28f23919/spikingjelly/clock_driven/neuron.py#L62

You can call  `self.register_memory('v',  some_value)` to set the initial states. Note that this vaule will also be used as the default reset value.

**Comment by asudeeaydin** [comment]
Thanks a lot, can I also initialize different neurons with different values using this?

**Comment by fangwei123456** [comment]
Yes, you can.

**Comment by asudeeaydin** [comment]
Thanks a lot! I really appreciate your quick responses. 

Another question, can we learn the threshold and membrane time constant values for LIF neurons? Or do we always need to manually select their value and have neurons with the same neuronal hyperparameters?

**Comment by fangwei123456** [comment]
> can we learn the threshold and membrane time constant values for LIF neurons?
Yes, you should delete the origin attribute and new a parameter like this:

```python
class LearnableLIFNode(neuron.LIFNode):
    def __init__(self, *args, **kwargs):
        super(LearnableLIFNode, self).__init__(*args, **kwargs)
        del self.v_threshold
        self.v_threshold = nn.Parameter(torch.as_tensor([1.]))
```

**Comment by asudeeaydin** [comment]
Is this different than the parametric LIF model? Can both the threshold and membrane potential be learnable?

Coming back to setting non-zero neuron states, if the initial states that I set will be the default reset values for the neurons, does this mean that after spiking neurons will reset to the states I've set? And the charge equation will change?

Is it possible to have the neurons reset back to 0 after spiking and have the charge equation as the classic LIF charge equation?

**Comment by fangwei123456** [comment]
> Is this different than the parametric LIF model? 

Yes, PLIF uses learnable tau but not learnable v_th.

> Can both the threshold and membrane potential be learnable?

You can set any attribute as learnable. But membrane potential `v` is determined by inputs and the neuron's parameters. You can not set `v` as learnable.


> does this mean that after spiking neurons will reset to the states I've set? And the charge equation will change?

Yes.

> Is it possible to have the neurons reset back to 0 after spiking and have the charge equation as the classic LIF charge equation?

I guess that you want `v_reset=0` and `v_init` is another user-defined value.

To implement this, you can change the charge function:

```python
class CustomVinitLIFNode(neuron.LIFNode):
    def __init__(self, v_init, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.v_init = v_init
        
    def neuronal_charge(self, x: torch.Tensor):
        if isinstance(self.v, float):
            # note that self.v == 0. after reset, and self.v becomes a tensor after any forward
            self.v = self.v_init
        super().neuronal_charge(x)
```

**Comment by asudeeaydin** [comment]
> You can set any attribute as learnable. But membrane potential v is determined by inputs and the neuron's parameters. You can not set v as learnable.

Yes, sorry, I meant to say membrane time constant. 

Do PLIF neurons learn the same tau layer-wise? So every channel in the same layer (for a spiking-CNN) would have the same tau? Is there a way to have the tau learnable channel-wise or neuron-wise? 

> I guess that you want `v_reset=0` and `v_init` is another user-defined value.

Yes exactly, thanks a lot.

**Comment by fangwei123456** [comment]
> Do PLIF neurons learn the same tau layer-wise? 

Yes.

>  So every channel in the same layer (for a spiking-CNN) would have the same tau?

Yes.

> Is there a way to have the tau learnable channel-wise or neuron-wise?

It can be implementd by setting this parameter with appropriate shape. Here is an example:

https://github.com/fangwei123456/spikingjelly/blob/940737e6f25687b1b93de41291603fa875e471f5/spikingjelly/activation_based/layer.py#L458

**Comment by asudeeaydin** [comment]
Sorry for jumping around topics, for the PLIF neuron, is this correct?

https://github.com/fangwei123456/spikingjelly/blob/cb2f0b9fb0659b1ddfa320ea83434cfa28f23919/spikingjelly/clock_driven/neuron.py#L604

shouldn't it be:
`tau = 1/self.w.sigmoid()`

I guess this `extra_repr()` function is just for accessing the value of tau? 

 For implementing channel-wise learnable membrane potentials, do you mean I would need to do something along the lines of: 
```
class ParametricLIFNode(BaseNode):
    def __init__(self, init_tau: float = 2.0, v_threshold: float = 1.,
                 v_reset: float = 0., surrogate_function: Callable = surrogate.Sigmoid(),
                 detach_reset: bool = False, shared_across_channels: bool = True, in_channels: int = 1):

        assert isinstance(init_tau, float) and init_tau > 1.
        super().__init__(v_threshold, v_reset, surrogate_function, detach_reset)

        self.shared_across_channels = shared_across_channels
        if self.shared_across_channels: # same membrane time constant accross channels of a layer
            init_w = - math.log(init_tau - 1.)
            self.w = nn.Parameter(torch.as_tensor(init_w))
        else: # channel-wise different membrane time constant 
            assert in_channels != 0
            init_w = - math.log(init_tau - 1.) * torch.ones(in_channels)
            self.w = nn.Parameter(init_w)

    def extra_repr(self):
        with torch.no_grad():
            tau = self.w.sigmoid()
        return super().extra_repr() + f', tau={tau}'

    def neuronal_charge(self, x: torch.Tensor):
        if self.shared_across_channels:
            if self.v_reset is None:
                self.v = self.v + (x - self.v) * self.w.sigmoid()
            else:
                if self.v_reset == 0.:
                    self.v = self.v + (x - self.v) * self.w.sigmoid()
                else:
                    self.v = self.v + (x - (self.v - self.v_reset)) * self.w.sigmoid()
        else: 
            if self.v_reset is None:
                self.v = self.v + (x - self.v) * self.w[:,None,None].sigmoid()
            else:
                if self.v_reset == 0.:
                    self.v = self.v + (x - self.v) * self.w[:,None,None].sigmoid()
                else:
                    self.v = self.v + (x - (self.v - self.v_reset)) * self.w[:,None,None].sigmoid()
```

**Comment by fangwei123456** [comment]
> shouldn't it be:
tau = 1/self.w.sigmoid()

Yes, I have fixed in new version:

https://github.com/fangwei123456/spikingjelly/blob/940737e6f25687b1b93de41291603fa875e471f5/spikingjelly/activation_based/neuron.py#L1069

> I guess this extra_repr() function is just for accessing the value of tau?

This function is used to print.

https://pytorch.org/docs/stable/generated/torch.nn.Module.html?highlight=extra_repr#torch.nn.Module.extra_repr


For your codes, I suggest that you set shape of the parameter as `[C, 1, 1]`, then it can be multiplied directly to inputs with shape `[*, C, H, W]` by boardcasting,

**Comment by asudeeaydin** [comment]
Oh, you mean like this?

```
class ParametricLIFNode(BaseNode):
    def __init__(self, init_tau: float = 2.0, v_threshold: float = 1.,
                 v_reset: float = 0., surrogate_function: Callable = surrogate.Sigmoid(),
                 detach_reset: bool = False, shared_across_channels: bool = True, in_channels: int = 0):

        assert isinstance(init_tau, float) and init_tau > 1.
        super().__init__(v_threshold, v_reset, surrogate_function, detach_reset)

        self.shared_across_channels = shared_across_channels
        if self.shared_across_channels: # same membrane time constant accross channels of a layer
            init_w = - math.log(init_tau - 1.)
            self.w = nn.Parameter(torch.as_tensor(init_w))
        else: # channel-wise different membrane time constant 
            assert in_channels != 0
            init_w = - math.log(init_tau - 1.) * torch.ones(in_channels, 1, 1)
            self.w = nn.Parameter(init_w)

    def extra_repr(self):
        with torch.no_grad():
            tau = 1/self.w.sigmoid()
        return super().extra_repr() + f', tau={tau}'

    def neuronal_charge(self, x: torch.Tensor):
        if self.v_reset is None:
            self.v = self.v + (x - self.v) * self.w.sigmoid()
        else:
            if self.v_reset == 0.:
                self.v = self.v + (x - self.v) * self.w.sigmoid()
            else:
                self.v = self.v + (x - (self.v - self.v_reset)) * self.w.sigmoid()
```

The shape of the parameters will be [C, 1, 1]?

**Comment by asudeeaydin** [comment]
Thank you, what you proposed worked. 

Additionally, are sigma-delta neurons supported in spiking jelly?

**Comment by asudeeaydin** [comment]
> I guess that you want `v_reset=0` and `v_init` is another user-defined value.
> 
> To implement this, you can change the charge function:
> 
> ```python
> class CustomVinitLIFNode(neuron.LIFNode):
>     def __init__(self, v_init, *args, **kwargs):
>         super().__init__(*args, **kwargs)
>         self.v_init = v_init
>         
>     def neuronal_charge(self, x: torch.Tensor):
>         if isinstance(self.v, float):
>             # note that self.v == 0. after reset, and self.v becomes a tensor after any forward
>             self.v = self.v_init
>         super().neuronal_charge(x)
> ```

After resetting the neuron potentials, I wouldn't be able to initialize the states again. I would like to initialize the states of neurons with different values in a script several times without having to call the init function and without changing the reset values and neuronal charge equations. In this case, just setting self.v = v_init work for what I want to do?

Would the following work?
```
class CustomVinitLIFNode(BaseNode):
    def __init__(self, tau: float = 2., v_threshold: float = 1.,
                 v_reset: float = 0., surrogate_function: Callable = surrogate.Sigmoid(),
                 detach_reset: bool = False):

        assert isinstance(tau, float) and tau > 1.

        super().__init__(v_threshold, v_reset, surrogate_function, detach_reset)
        self.tau = tau
        
    def neuronal_charge(self, x: torch.Tensor, v_init: torch.Tensor):
            # note that self.v == 0. after reset, and self.v becomes a tensor after any forward
        if v_init != None: # If a v_init is passed in 
            self.v = v_init

        if self.v_reset is None:
            self.v = self.v + (x - self.v) / self.tau

        else:
            if isinstance(self.v_reset, float) and self.v_reset == 0.:
                self.v = self.v + (x - self.v) / self.tau
            else: # charge equation
                self.v = self.v + (x - (self.v - self.v_reset)) / self.tau
    
    def forward(self, x: torch.Tensor, v_init: torch.Tensor):
        self.neuronal_charge(x, v_init)
        self.neuronal_fire()
        self.neuronal_reset()
        return self.spike
```

**Comment by fangwei123456** [comment]
Yes, according to your codes, you can set an arbitrary `v_init` when you call `forward` with `v_init is not None`.
